### ISSUE TITLE
Optimize quadtree rebalancing

### ISSUE THREAD

**Issue by parasyte** [melonjs/melonJS#551]
From discussion in #544: The quadtree is cleared and rebuilt on each frame. It can be optimized by allowing the objects to reposition themselves within the tree as they move about the world. Performance gain will be minimal, but may impact mobile more than desktop...

**Comment by obiot** [comment]
+1 for that one that felt behind ! we should probably start by adding a test unit for the quadtree implementation. That will help ensuring we don't break everything then :):):)

**Comment by obiot** [comment]
i just realize i completely overlooked this thing (although yes, and yes is was written in the initial message of this ticket) . Remove an item is actually the easy part, but to make it work, the whole tree also needs to be "re-validated" every time an object is moving, which makes me wonder if this is not almost the same (in terms of required cpu-cycle) than just rebuild the tree from scratch on every frame like we do now.

basically it could be something like : check if every object of the tree are still i their initial node (or add a way for an object to notify the quadtree it changed position), if not, remove the object, clean the node/sub-node if necessary, and the re-insert it.

**Comment by obiot** [comment]
backported the changes from the #551 branch to master, and changing the milestone to "future" (4.1.0?)

**Comment by parasyte** [comment]
FWIW, you don't have to rebalance the tree for removals, just adds.

**Comment by obiot** [comment]
yes should be something like this actually. To be honest the biggest part was to add this "remove" function, but i just don't have the time (if we want to have a 4.0 version released this year) to fully debug it (although it seems to be working pretty well now) and add the missing part of this ticket.
